### PR TITLE
Add wi4mpi recipe

### PR DESCRIPTION
**Pull request by marc-joos**
This PR adds `wi4mpi` package; WI4MPI is a Wrapper Interface For MPI performing a light translation between MPI constants and MPI objects from an MPI implementation to another one.

### CODE REVIEW

File: var/spack/repos/builtin/packages/wi4mpi/package.py [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+            env.set('WI4MPI_CXX', "icpc")
+            env.set('WI4MPI_FC', "ifort")
+        elif '%clang' in self.spec:
+            env.set('WI4MPI_CC', "clang")
+            env.set('WI4MPI_CXX', "clang++")
+            env.set('WI4MPI_FC', "flang")
+        elif '%pgi' in self.spec:
+            env.set('WI4MPI_CC', "pgcc")
+            env.set('WI4MPI_CXX', "pgc++")
+            env.set('WI4MPI_FC', "pgfortran")
```
Comment (alalazo) [inline var/spack/repos/builtin/packages/wi4mpi/package.py]: Do you support mixed toolchains? I think this should report the compilers in `compilers.yaml`

File: var/spack/repos/builtin/packages/wi4mpi/package.py [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+        elif '%pgi' in self.spec:
+            compiler = "PGI"
+        else:
+            tty.error("Could not determine compiler used")
+        wi4mpi_build_type = 'RELEASE'
+        if self.spec.variants["build_type"].value == "RelWithDebInfo":
+            wi4mpi_build_type = 'NORMAL'
+        elif self.spec.variants["build_type"].value == "Debug":
+            wi4mpi_build_type = 'DEBUG'
+        args = ["-DWI4MPI_REALEASE=" + wi4mpi_build_type,
```
Comment (alalazo) [inline var/spack/repos/builtin/packages/wi4mpi/package.py]: Maybe a typo:
```suggestion
        args = ["-DWI4MPI_RELEASE=" + wi4mpi_build_type,

```
?

Comment (alalazo) [reply var/spack/repos/builtin/packages/wi4mpi/package.py]: Also note that a more idiomatic way to write this is:
```python
args = [
    self.define('WI4MPI_RELEASE', wi4mpi_build_type),
    self.define('WI4MPI_COMPILER', compiler)
]
```

**Review by alalazo — CHANGES_REQUESTED** [review]
I left a few requests / questions. Let me know what you think.

File: var/spack/repos/builtin/packages/wi4mpi/package.py (hunk 1 above)
Comment (marc-joos) [reply var/spack/repos/builtin/packages/wi4mpi/package.py]: We do not support mixed toolchains at the moment

File: var/spack/repos/builtin/packages/wi4mpi/package.py (hunk 2 above)
Comment (marc-joos) [reply var/spack/repos/builtin/packages/wi4mpi/package.py]: It is indeed a typo but in Wi4MPI base code itself. We will fix it in a future release and change it accordingly in Spack recipe.

Thank you for the suggestion to improve argument passing; I've just pushed a new commit with the suggested edit.

File: var/spack/repos/builtin/packages/wi4mpi/package.py (hunk 1 above)
Comment (alalazo) [reply var/spack/repos/builtin/packages/wi4mpi/package.py]: I think you can avoid all these conditionals and just do:
```python
env.set('WI4MPI_CC', self.compiler.cc)
env.set('WI4MPI_CXX', self.compiler.cxx)
env.set('WI4MPI_FC', self.compiler.fc)

```

**Review by alalazo — CHANGES_REQUESTED** [review]
Another minor suggestion based on previous feedback. Let me know if that works.

File: var/spack/repos/builtin/packages/wi4mpi/package.py (hunk 1 above)
Comment (marc-joos) [reply var/spack/repos/builtin/packages/wi4mpi/package.py]: Thank you for the suggestion, we modified the code accordingly

**Review by alalazo — APPROVED** [review]

**Pull request merged**